sycl/source/device_selector.cpp

```diff
@@ -15,76 +15,133 @@
 
 __SYCL_INLINE_NAMESPACE(cl) {
 namespace sycl {
+
+// Utility function to check if device is of the preferred SYCL_BE.
+static bool isDeviceOfPreferredSyclBe(const device &Device) {
+  // Taking the version information from the platform gives us more useful
+  // information than the driver_version of the device.
+  const platform Platform = Device.get_info<info::device::platform>();
+  const std::string PlatformVersion =
+      Platform.get_info<info::platform::version>();
+
+  detail::pi::Backend PreferredBE = detail::pi::getPreferredBE();
+
+  if (PreferredBE == RT::Backend::SYCL_BE_PI_OPENCL) {
+    if (PlatformVersion.find("OpenCL") != std::string::npos) {
+      return true;
+    }
+  } else if (PreferredBE == RT::Backend::SYCL_BE_PI_CUDA) {
+    if (PlatformVersion.find("CUDA") != std::string::npos) {
+      return true;
+    }
+  }
+  return false;
+}
+
 device device_selector::select_device() const {
   vector_class<device> devices = device::get_devices();
   int score = -1;
   const device *res = nullptr;
-  for (const auto &dev : devices)
-    if (score < operator()(dev)) {
+  for (const auto &dev : devices) {
+    int dev_score = operator()(dev);
+    if (detail::pi::trace(detail::pi::TraceLevel::PI_TRACE_ALL)) {
+      std::cout << "SYCL_PI_TRACE[1]: select_device(): -> score = " << score
```

Comment:
It doesn't sound like pi trace. It is pure SYCL thing.

Reply:
This is a very useful info, suggest with which trace it should be printed. I also don't see anything bad that verbose PI tracing wants to print something that is "pure SYCL".

Reply:
Probably, we need to introduce some non-PI tracing eventually. Since it is the only one example of where it is needed I'm OK with keeping things as is.